<a href="https://colab.research.google.com/github/mohandaz/HIA-302-Project/blob/main/HIA302_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

CLone Github Repository : https://github.com/mohandaz/HIA-302-Project

In [1]:
!git clone https://github.com/mohandaz/HIA-302-Project

Cloning into 'HIA-302-Project'...
remote: Enumerating objects: 50, done.
remote: Counting objects: 100% (50/50), done.
remote: Compressing objects: 100% (46/46), done.
remote: Total 50 (delta 22), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (50/50), 280.90 KiB | 9.06 MiB/s, done.
Resolving deltas: 100% (22/22), done.


Libraries used

In [81]:
import pandas as pd
import csv
import numpy as np

2.1 Demonstrate different ways to convert the horse-colic.txt file to horse-colic.csv.

METHOD 1 : Using Pandas to read and write

In [82]:
import pandas as pd

# Input and output file paths
input_dir = '/content/HIA-302-Project/horse-colic.txt'
output_dir = '/content/HIA-302-Project/horse-colic.csv'

# Read the TXT file with a preexisting index and header
df = pd.read_csv(input_dir, header=None)

# Save the DataFrame to a new CSV file
df.to_csv(output_dir)

print(f"Conversion completed. CSV file saved at: {output_dir}")


Conversion completed. CSV file saved at: /content/HIA-302-Project/horse-colic.csv


METHOD 2: Using CSV to read and Pandas to write

In [27]:
import csv
import pandas as pd

# Input and output file paths
input_file = '/content/HIA-302-Project/horse-colic.txt'
output_file = '/content/HIA-302-Project/horse-colic.csv'

# Read data using csv module
with open(input_file, 'r') as csv_file:
    csv_reader = csv.reader(csv_file)
    data = list(csv_reader)

# Create a DataFrame using pandas
df = pd.DataFrame(data, columns=None)

# Write to CSV file with empty header and default index
df.to_csv(output_file)

print(f"Conversion completed. CSV file saved at: {output_dir}")

Question 2.2
 : Demonstrate that the csv file can read in Jupyter Notebook, and the dataset remains the same as the raw dataset.


In [ ]:
# Specify the path to the CSV file
output_dir = '/content/HIA-302-Project/horse-colic.csv'

# Read the CSV file into a Pandas DataFrame
horse_colic = pd.read_csv(output_dir, index_col=0, header=0)

# Display the first few rows of the CSV DataFrame

horse_colic.info()
horse_colic.head()


2.3 Replacing the missing values (i.e. with"?" with NaN (Not a Number) value in a loaded dataset using python

In [98]:
horse_colic_with_NaN = horse_colic.replace("?","NaN")
horse_colic_with_NaN.head()


0  1        2      3    4   5    6    7    8  9  ...     18    19   20  \
0  2  1   530101  38.50   66  28    3    3  NaN  2  ...  45.00  8.40  NaN   
1  1  1   534817   39.2   88  20  NaN  NaN    4  1  ...     50    85    2   
2  2  1   530334  38.30   40  24    1    1    3  1  ...  33.00  6.70  NaN   
3  1  9  5290409  39.10  164  84    4    1    6  2  ...  48.00  7.20    3   
4  2  1   530255  37.30  104  35  NaN  NaN    6  2  ...  74.00  7.40  NaN   

     21 22 23     24 25 26 27  
0   NaN  2  2  11300  0  0  2  
1     2  3  2   2208  0  0  2  
2   NaN  1  2      0  0  0  1  
3  5.30  2  1   2208  0  0  1  
4   NaN  2  2   4300  0  0  2  

[5 rows x 28 columns]

2.4 Save the latest dataset with NaN as horse_colic_with_Nan.csv and ensure the data frame's row number is not saved into csv file

In [100]:
output_dir = '/content/HIA-302-Project/horse_colic_with_NaN.csv'
# Save the DataFrame to a CSV file
horse_colic_with_NaN.to_csv(output_dir, index=False)

print(f"CSV file saved at: {output_dir}")

horse_colic_with_NaN.info()

CSV file saved at: /content/HIA-302-Project/horse_colic_with_NaN.csv
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Surgery      299 non-null    float64
 1   Age          300 non-null    int64  
 2   Hosp_No      300 non-null    int64  
 3   Rect_Temp    240 non-null    float64
 4   Pulse        276 non-null    float64
 5   RR           242 non-null    float64
 6   Temp_Ext     244 non-null    float64
 7   Peri_pulse   231 non-null    float64
 8   Muc_Memb     253 non-null    float64
 9   CRT          268 non-null    float64
 10  Pain         245 non-null    float64
 11  Perist       256 non-null    float64
 12  Abd_Distn    244 non-null    float64
 13  NGT          196 non-null    float64
 14  Naso_Refx    194 non-null    float64
 15  NRP          53 non-null     float64
 16  Rect_Exm     198 non-null    float64
 17  Abd          182 non-nu

2.5 Rename the header by following the description file (horse-colic.names) with not more than 10 characters i the header title

In [103]:
import pandas as pd
import os

# Specify input and output file paths
input_dir = '/content/HIA-302-Project/horse_colic_with_NaN.csv'
output_dir = '/content/HIA-302-Project/horse-colic-with-title.csv'

# Read the DataFrame from the input CSV file
df = pd.read_csv(input_dir)

# Reset the index of the DataFrame
df.reset_index(drop=True, inplace=True)

# Define columns that should be integers
integer_columns = ['Surgery', 'Age', 'Hosp_No', 'Surg_Lesi', 'Lesi_Site', 'Lesi_Type', 'Lesi_Sub', 'Cp_Data']

# Convert specified columns to integers
df[integer_columns] = df[integer_columns].apply(pd.to_numeric, errors='coerce').astype('Int64')

# Save the DataFrame to a CSV file
df.to_csv(output_dir, index=False)

# Print a message indicating that the CSV file has been saved
print(f"CSV file saved at: {output_dir}")


CSV file saved at: /content/HIA-302-Project/horse-colic-with-title.csv


In [104]:
df.head()

Surgery  Age  Hosp_No  Rect_Temp  Pulse    RR  Temp_Ext  Peri_pulse  \
0        2    1   530101       38.5   66.0  28.0       3.0         3.0   
1        1    1   534817       39.2   88.0  20.0       NaN         NaN   
2        2    1   530334       38.3   40.0  24.0       1.0         1.0   
3        1    9  5290409       39.1  164.0  84.0       4.0         1.0   
4        2    1   530255       37.3  104.0  35.0       NaN         NaN   

   Muc_Memb  CRT  ...   PCV  Tot_Prot  Abc_App  Abc_Protein  Outcome  \
0       NaN  2.0  ...  45.0       8.4      NaN          NaN      2.0   
1       4.0  1.0  ...  50.0      85.0      2.0          2.0      3.0   
2       3.0  1.0  ...  33.0       6.7      NaN          NaN      1.0   
3       6.0  2.0  ...  48.0       7.2      3.0          5.3      2.0   
4       6.0  2.0  ...  74.0       7.4      NaN          NaN      2.0   

   Surg_Lesi  Lesi_Site  Lesi_Type  Lesi_Sub  Cp_Data  
0          2      11300          0         0        2  
1          2       2208          0         0        2  
2          2          0          0         0        1  
3          1       2208          0         0        1  
4          2       4300          0         0        2  

[5 rows x 28 columns]

2.6 Save the latest dataset with header title as horse_colic_with_title.csv and ensure the data frame's row number is not saved into csv file.

2.7 Load the horse_colic_with_title.csv file and display list of total NaN at each column in the data set.

In [113]:
nan_counts = df.isna().sum()
nan_counts_total = df.isna().sum().sum()
# Display the total number of NaN values in the DataFrame
print("Total NaN values in the DataFrame:", total_nan_count)

# Display the total number of NaN values for each column
print("Total NaN values in each column:")
print(nan_counts)
print(nan_counts_total)

Total NaN values in the DataFrame: 1605
Total NaN values in each column:
Surgery          1
Age              0
Hosp_No          0
Rect_Temp       60
Pulse           24
RR              58
Temp_Ext        56
Peri_pulse      69
Muc_Memb        47
CRT             32
Pain            55
Perist          44
Abd_Distn       56
NGT            104
Naso_Refx      106
NRP            247
Rect_Exm       102
Abd            118
PCV             29
Tot_Prot        33
Abc_App        165
Abc_Protein    198
Outcome          1
Surg_Lesi        0
Lesi_Site        0
Lesi_Type        0
Lesi_Sub         0
Cp_Data          0
dtype: int64
1605


2.8 Visualize the totsl NaN at each column in the dataset with a bar chart and set colour to 'blue' using Metplotlib library

2.9 Retrive the datatypes for the dataset

2.10 In this dataset, will it be good idea to delete/drop tha NaN rows? Justify your answer

2.11 Read and evaluate the following codes in Figure 1. What method is the code using to reduce missing values from 1605 to 0? Justify your answer

2.12 Fill out the inline comments marked with hashtags # in lines 8,11,17,20,22. and 26. include the missing codes in line 5 and 6